relay-wstring/src/lib.rs

```diff
@@ -0,0 +1,512 @@
+
+impl Utf16Error {
+    /// Create a new [Utf16Error].
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for Utf16Error {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A UTF-16 [str]-like type with little-endian byte order.
+///
+/// This mostly behaves like [str] does for UTF-8 encoded bytes slices, but works with
+/// UTF-16LE encoded byte slices.
+///
+/// See the [module-level documentation](index.html) for some simple examples.
+#[derive(Debug, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct WStr {
+    raw: [u8],
+}
+
+impl WStr {
+    /// Create a new `&WStr` from an existing UTF-16 little-endian encoded byte-slice.
+    ///
+    /// If the byte-slice is not valid [Utf16Error] is returned.
+    pub fn from_utf16le(raw: &[u8]) -> Result<&Self, Utf16Error> {
+        validate_raw_utf16le(raw)?;
+        Ok(unsafe { Self::from_utf16le_unchecked(raw) })
+    }
+
+    /// Create a new `&mut WStr` from an existing UTF-16 little-endian encoded byte-slice.
+    ///
+    /// If the byte-slice is not valid [Utf16Error] is returned.
+    pub fn from_utf16le_mut(raw: &mut [u8]) -> Result<&mut Self, Utf16Error> {
+        validate_raw_utf16le(raw)?;
+        Ok(unsafe { Self::from_utf16le_unchecked_mut(raw) })
+    }
+
+    /// Create a new [WStr] from an existing UTF-16 little-endian encoded byte-slice.
+    ///
+    /// # Safety
+    ///
+    /// You must guarantee that the buffer passed in is encoded correctly otherwise you will
+    /// get undefined behaviour.
+    pub unsafe fn from_utf16le_unchecked(raw: &[u8]) -> &Self {
+        &*(raw as *const [u8] as *const Self)
+    }
+
+    /// Like [Self::from_utf16le_unchecked] but return a mutable reference.
+    ///
+    /// # Safety
+    ///
+    /// You must guarantee that the buffer passed in is encoded correctly otherwise you will
+    /// get undefined behaviour.
+    pub unsafe fn from_utf16le_unchecked_mut(raw: &mut [u8]) -> &mut Self {
+        &mut *(raw as *mut [u8] as *mut Self)
+    }
+
+    /// The length in bytes, not chars or graphemes.
+    pub fn len(&self) -> usize {
+        self.raw.len()
+    }
+
+    /// Returns `true` if the [Self::len] is zero.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns `true` if the index into the bytes is on a char boundary.
+    pub fn is_char_boundary(&self, index: usize) -> bool {
+        if index == 0 || index == self.len() {
+            return true;
+        }
+        if index % 2 != 0 {
+            return false;
+        }
+
+        // Since we always have a valid UTF-16LE string in here we now are sure we always
+        // have a byte at index + 1.  The only invalid thing now is a trailing surrogate.
+
+        let mut buf: [u8; 2] = [0; 2]; // TODO: avoid init
+        buf.copy_from_slice(&self.raw[index..index + 2]);
+        let u = u16::from_le_bytes(buf);
+        u & 0xDC00 != 0xDC00
+    }
+
+    /// Convert to a byte slice.
```

Comment:
_nit_: Doc comments should be in third person, like the ones above. For example "Converts" rather than "Convert". This also applies to the functions below.

Example: https://doc.rust-lang.org/stable/std/primitive.str.html#method.as_ptr